test/distributed/test_ddp_with_rpc.py

```diff
@@ -0,0 +1,444 @@
+    examples_per_trainer = int(n / NUM_TRAINERS)
+    return [
+        FeatureSet(
+            dense_features=training_examples.dense_features[
+                start : start + examples_per_trainer, :
+            ],
+            sparse_features=training_examples.sparse_features[
+                start : start + examples_per_trainer
+            ],
+            labels=training_examples.labels[
+                start : start + examples_per_trainer
+            ],
+        )
+        for start in range(0, n, examples_per_trainer)
+    ]
+
+
+class TestDdpWithRpc(MultiProcessTestCase):
+    rpc_backend = rpc.backend_registry.BackendType.PROCESS_GROUP
+    rpc_backend_options = None
+
+    @property
+    def world_size(self) -> int:
+        return WORLD_SIZE
+
+    def remote_worker_name(self) -> str:
+        # The name has to be consistent with that in 'dist_init' decorator.
+        return f"worker{REMOTE_WORKER_RANK}"
+
+    def trainer_name(self, rank):
+        # The name has to be consistent with that in 'dist_init' decorator.
+        return f"worker{rank}"
+
+    def setUp(self):
+        super(TestDdpWithRpc, self).setUp()
+
+        os.environ["MASTER_ADDR"] = str(MASTER_ADDR)
+        os.environ["MASTER_PORT"] = str(MASTER_PORT)
+        self._spawn_processes()
+
+    def tearDown(self):
+        super(TestDdpWithRpc, self).tearDown()
+
+    @dist_init
+    def _remote_worker_process(self):
+        process_group_for_ddp = dist_c10d.new_group(ranks=TRAINER_RANKS)
+        gLogger.info(f"The remote worker is running.")
+        dist.destroy_process_group(process_group_for_ddp)
+        gLogger.info(f"Exiting remote worker.")
+
+    @dist_init
+    def _trainer_process(self, rank: int):
+        gLogger.info(f"Running the trainer #{rank}...")
+
+    def do_test_on_master(
+        self,
+        ddp_mode: DdpMode,
+        trainer_method: Callable[[FeatureSet], None],
+        remote_em_rref: rpc.RRef,
+        remote_net_rref: rpc.RRef,
+    ):
+        trainer_rrefs = []
+        for rank in TRAINER_RANKS:
+            trainer = self.trainer_name(rank)
+            trainer_rrefs.append(
+                rpc.remote(
+                    trainer,
+                    Trainer,
+                    args=(remote_em_rref, remote_net_rref, ddp_mode, rank),
+                )
+            )
+
+        training_examples = get_training_examples()
+        for epoch in range(NUM_EPOCH):
+            futures = []
+            for idx, trainer_rref in enumerate(trainer_rrefs):
+                futures.append(
+                    _remote_method_async(
+                        trainer_method, trainer_rref, training_examples[idx]
+                    )
+                )
+            log_loss = 0
+            for future in futures:
+                log_loss += future.wait()
+            gLogger.info(
+                f"Log loss at epoch #{epoch}: {log_loss / NUM_TRAINERS}"
+            )
+
+    @dist_init
```

Comment:
Shouldn't dist_init wrap the individual test functions instead of these methods?